src/Components/ResourcePicker/ResourcePicker.tsx

```diff
@@ -0,0 +1,649 @@
+                            (option) =>
+                                option.itemType ===
+                                    SelectableOptionMenuItemType.Header &&
+                                option.text === freeformOptionsHeaderText
+                        );
+                        newOptions = freeFromOptionsHeaderExist
+                            ? [...newOptions, newOption]
+                            : [...newOptions, freeformOptionsHeader, newOption];
+                        setOptions(newOptions);
+                        setSelectedOption(newOption);
+                        setSelectedKey(newParsedOptionValue);
+                    }
+                    if (onResourceChange)
+                        onResourceChange(
+                            newParsedOptionValue,
+                            newOptions
+                                ?.filter(
+                                    (option) =>
+                                        option.itemType !==
+                                        SelectableOptionMenuItemType.Header
+                                )
+                                .map((option) => option.data || option.text) ||
+                                []
+                        );
+                } else {
+                    setSelectedOption(existingOption);
+                    setSelectedKey(existingOption.key);
+                    if (onResourceChange)
+                        onResourceChange(
+                            existingOption.data || existingOption.text,
+                            options
+                                ?.filter(
+                                    (option) =>
+                                        option.itemType !==
+                                        SelectableOptionMenuItemType.Header
+                                )
+                                .map((option) => option.data || option.text) ||
+                                []
+                        );
+                }
+            }
+        },
+        [options]
+    );
+
+    const handleOnRemove = useCallback(
+        (option: IComboBoxOption) => {
+            const restOfOptions = options.filter((o) => o.key !== option.key);
+            if (restOfOptions.length) {
+                const lastOption = restOfOptions[restOfOptions.length - 1];
+                if (
+                    lastOption.itemType === SelectableOptionMenuItemType.Header
+                ) {
+                    // remove the '---' freeform options header if there is no option below
+                    restOfOptions.splice(-1);
+                }
+            }
+            setOptions(restOfOptions);
+
+            if (option.key === selectedOption?.key) {
+                setSelectedOption(null);
+                if (onResourceChange)
+                    onResourceChange(
+                        null,
+                        restOfOptions
+                            .filter(
```

Comment:
we seem to do this filtering a lot, can we pull it out to a helper of some sort to simplify? or have a wrapper function around the onResourceChange that all the places call so they don't have to filter out the headers in every spot?